### PR TITLE
Enforcing [Event] attribute on publish calls.

### PR DESCRIPTION
**Pull request by JeffAshton**

### CODE REVIEW

File: src/D2L.CodeStyle.Analyzers/ApiUsage/Events/EventPublisherEVentTypesAnalyzer.cs [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+			bool hasEventAttr = eventTypeSymbol
+				.GetAttributes()
+				.Any( attr => attr.AttributeClass.Equals( eventAttributeType ) );
+
+			if( hasEventAttr ) {
+				return;
+			}
+
+			if( eventTypeSymbol.TypeKind == TypeKind.TypeParameter ) {
```
Comment (JeffAshton) [inline src/D2L.CodeStyle.Analyzers/ApiUsage/Events/EventPublisherEVentTypesAnalyzer.cs]: I'm going to take a second pass at these.  A few happened with wrappers:

http://search.dev.d2l/source/xref/Lms/lp/framework/core/D2L.LP/LP/Distributed/Events/Domain/Default/CompositeEventNotifier.cs

File: src/D2L.CodeStyle.Analyzers/ApiUsage/Events/EventPublisherEVentTypesAnalyzer.cs [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+			IImmutableSet<ISymbol> genericPublishMethods = Enumerable
+				.Concat(
+					GetGenericPublishMethods( compilation, IEventNotifierFullName ),
+					GetGenericPublishMethods( compilation, IEventPublisherFullName )
+				)
+				.ToImmutableHashSet();
+
+			context.RegisterSyntaxNodeAction(
+					ctxt => {
+						if( ctxt.Node is InvocationExpressionSyntax invocation ) {
```
Comment (j3parker) [inline src/D2L.CodeStyle.Analyzers/ApiUsage/Events/EventPublisherEVentTypesAnalyzer.cs]: I don't think this guard is necessary. Pretty sure I checked this in Roslyn's source. I just assume the cast passes when I write analyzers.

Comment (j3parker) [reply src/D2L.CodeStyle.Analyzers/ApiUsage/Events/EventPublisherEVentTypesAnalyzer.cs]: i.e. if you want to pull it out of the context for your analyze method this works:

```cs
context.RegisterSyntaxNode(
  ctx => AnalyzeMethodInvocation(
    ctx,
    ctx.Node,
    eventAttributeType,
    genericPublishMethods
  )
);
```

Comment (j3parker) [reply src/D2L.CodeStyle.Analyzers/ApiUsage/Events/EventPublisherEVentTypesAnalyzer.cs]: - [ ] remove unnecessary guard

Comment (JeffAshton) [reply src/D2L.CodeStyle.Analyzers/ApiUsage/Events/EventPublisherEVentTypesAnalyzer.cs]: Switched to a cast.  I'd rather an invalid cast exception then a null reference exception.

Comment (j3parker) [reply src/D2L.CodeStyle.Analyzers/ApiUsage/Events/EventPublisherEVentTypesAnalyzer.cs]: Don't worry, its just that they couldn't get the API they really wanted in C#. The cast will never fail.

Comment (j3parker) [reply src/D2L.CodeStyle.Analyzers/ApiUsage/Events/EventPublisherEVentTypesAnalyzer.cs]: But either way is good, ya hard casting it is probably what I'd do too on second thought

**Pull request merged**